src/Common/src/CoreLib/System/Memory.cs

```diff
@@ -43,7 +47,7 @@ public Memory(T[] array)
             if (array == null)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
             if (default(T) == null && array.GetType() != typeof(T[]))
-                ThrowHelper.ThrowArrayTypeMismatchException();
+                ThrowHelper.ThrowArrayTypeMismatchException_ArrayTypeMustBeExactMatch(typeof(T));
```

Comment:
Passing in typeof(T) has non-trivial overhead here. I do not think it is worth doing for this rare exception.